vaporetto/src/type_scorer.rs

```diff
@@ -0,0 +1,205 @@
+use crate::sentence::Sentence;
+use daachorse::DoubleArrayAhoCorasick;
+
+pub enum TypeScorer {
+    Pma(TypeScorerPma),
+    Cache(TypeScorerCache),
+}
+
+impl TypeScorer {
+    pub fn new(
+        pma: DoubleArrayAhoCorasick,
+        weights: Vec<Vec<ScoreValue>>,
+        window_size: usize,
+    ) -> Self {
+        if window_size <= 3 {
+            Self::Cache(TypeScorerCache::new(pma, weights, window_size))
+        } else {
+            Self::Pma(TypeScorerPma::new(pma, weights, window_size))
+        }
+    }
+
+    pub fn add_scores(&self, sentence: &Sentence, start: usize, ys: &mut [ScoreValue]) {
+        match self {
+            TypeScorer::Pma(pma) => pma.add_scores(sentence, start, ys),
+            TypeScorer::Cache(cache) => cache.add_scores(sentence, start, ys),
+        }
+    }
+}
+
+pub struct TypeScorerPma {
+    pma: DoubleArrayAhoCorasick,
+    weights: Vec<Vec<ScoreValue>>,
+    window_size: usize,
+}
+
+impl TypeScorerPma {
+    pub fn new(
+        pma: DoubleArrayAhoCorasick,
+        weights: Vec<Vec<ScoreValue>>,
+        window_size: usize,
+    ) -> Self {
+        Self {
+            pma,
+            weights,
+            window_size,
+        }
+    }
+
+    pub fn add_scores(&self, sentence: &Sentence, start: usize, ys: &mut [ScoreValue]) {
+        let type_start = if start >= self.window_size {
+            start + 1 - self.window_size
+        } else {
+            0
+        };
+        let type_end = std::cmp::min(
+            start + ys.len() + self.window_size,
+            sentence.char_type.len(),
+        );
+        let char_type = &sentence.char_type[type_start..type_end];
+        let padding = start - type_start + 1;
+        for m in self.pma.find_overlapping_no_suffix_iter(&char_type) {
+            let offset = m.end() as isize - self.window_size as isize - padding as isize;
+            let weights = &self.weights[m.pattern()];
+            if offset >= 0 {
+                for (w, y) in weights.iter().zip(&mut ys[offset as usize..]) {
+                    *y += w;
+                }
+            } else {
+                for (w, y) in weights[-offset as usize..].iter().zip(ys.iter_mut()) {
+                    *y += w;
+                }
+            }
+        }
+    }
+}
+
+pub struct TypeScorerCache {
+    scores: Vec<ScoreValue>,
+    window_size: usize,
+    sequence_mask: usize,
+}
+
+impl TypeScorerCache {
+    pub fn new(
+        pma: DoubleArrayAhoCorasick,
+        weights: Vec<Vec<ScoreValue>>,
+        window_size: usize,
+    ) -> Self {
+        let sequence_size = window_size * 2;
+        let all_sequences = ALPHABET_SIZE.pow(sequence_size as u32);
+
+        let mut sequence = vec![0u8; sequence_size];
+        let mut scores = vec![0 as ScoreValue; all_sequences];
+
+        for (i, score) in scores.iter_mut().enumerate() {
+            if !Self::seqid_to_seq(i, &mut sequence) {
+                continue;
+            }
+            let mut y = 0;
```

Comment:
Could you fix errors on the CI?
```
$ cargo test -p vaporetto --no-default-features
...
error[E0277]: cannot add-assign `f64` to `{integer}`
   --> vaporetto/src/type_predictor.rs:102:19
    |
102 |                 y += weights[m.pattern()][sequence_size - m.end()];
    |                   ^^ no implementation for `{integer} += f64`
    |
    = help: the trait `AddAssign<f64>` is not implemented for `{integer}`

error[E0308]: mismatched types
   --> vaporetto/src/type_predictor.rs:104:22
    |
104 |             *score = y;
    |                      ^ expected `f64`, found integer
```

```suggestion
            let mut y = ScoreValue::default();
```

Reply:
Fixed (I'm sorry I forgot to use Commit suggestion)